community/detectors/gitlab_cve_2021_22205/src/main/java/com/google/tsunami/plugins/detectors/cves/cve202122205/Cve202122205VulnDetector.java

```diff
@@ -0,0 +1,269 @@
+    return targetUrlBuilder.toString();
+  }
+
+  private static String buildTargetVulnerabilityUrl(NetworkService networkService) {
+    StringBuilder targetUrlBuilder = new StringBuilder();
+    if (NetworkServiceUtils.isWebService(networkService)) {
+      targetUrlBuilder.append(NetworkServiceUtils.buildWebApplicationRootUrl(networkService));
+    } else {
+      targetUrlBuilder
+          .append("http://")
+          .append(toUriAuthority(networkService.getNetworkEndpoint()))
+          .append("/");
+    }
+    targetUrlBuilder.append(VUL_PATH);
+    return targetUrlBuilder.toString();
+  }
+
+  @Override
+  public DetectionReportList detect(
+      TargetInfo targetInfo, ImmutableList<NetworkService> matchedServices) {
+    logger.atInfo().log("CVE-2921-22205 starts detecting.");
+
+    return DetectionReportList.newBuilder()
+        .addAllDetectionReports(
+            matchedServices.stream()
+                .filter(Cve202122205VulnDetector::isWebServiceOrUnknownService)
+                .filter(this::isServiceVulnerable)
+                .map(networkService -> buildDetectionReport(targetInfo, networkService))
+                .collect(toImmutableList()))
+        .build();
+  }
+
+  private String clearCookie(String cookie) {
+    return cookie.split("path=/;")[0];
+  }
+
+  private Map<String, String> getCsrfTokenAndCookie(NetworkService networkService) {
+    String targetUserSignUrl = buildTargetUserSignUrl(networkService);
+    HashMap<String, String> result = new HashMap<>();
+    try {
+      HttpResponse httpResponse =
+          httpClient.send(get(targetUserSignUrl).withEmptyHeaders().build(), networkService);
+      if (httpResponse.status().code() == 200) {
+        String cookie = httpResponse.headers().get(SET_COOKIE).get();
+        Matcher csrfTokenMatcher = CSRF_TOKEN_PATTERN.matcher(httpResponse.bodyString().get());
+        if (csrfTokenMatcher.find() && !cookie.isEmpty()) {
+          result.put(CSRF_TOKEN, csrfTokenMatcher.group(1));
+          StringBuilder cookies = new StringBuilder();
+          Iterator<String> headerCookies =
+              httpResponse.headers().getAll(SET_COOKIE).stream().iterator();
+          while (headerCookies.hasNext()) {
+            cookies.append(clearCookie(headerCookies.next()));
+          }
+          result.put(COOKIE, cookies.toString());
+        }
+      }
+    } catch (IOException e) {
+      logger.atWarning().withCause(e).log("Request to target %s failed", networkService);
+      return null;
+    }
+    return result;
+  }
+
+  private boolean isServiceVulnerable(NetworkService networkService) {
+    String targetVulnerabilityUrl = buildTargetVulnerabilityUrl(networkService);
+    Map<String, String> preInfo = getCsrfTokenAndCookie(networkService);
+    if (Objects.requireNonNull(preInfo).isEmpty()) {
+      logger.atWarning().log("Get %s csrf_token failed", networkService);
+      return false;
+    }
+    try {
+      byte[] payload = Hex.decodeHex(POST_DATA);
```

Comment:
Please use Guava code for hex decoding instead, i.e. `byte[] payload = BaseEncoding.base16().decode(POST_DATA);`. You have to convert all the lowercase characters in `POST_DATA` to uppercase first.